### ISSUE TITLE
Promote Redox to Tier 2 Platform

### ISSUE THREAD

**Issue by jackpot51** [rust-lang/rust#43206]
Redox is unique among targets. It has libstd support down to the syscall level, completely in Rust. For a while, it has been a Tier 3 platform (https://forge.rust-lang.org/platform-support.html). Due to recent advances in Redox, I think it is now time to promote it to a Tier 2 platform.

What this would entail is the merging of this changeset: https://github.com/rust-lang/rust/compare/master...redox-os:master

To summarize, this adds a docker CI image for Redox, adds the CI image to Travis, and fixes any remaining compilation problems with Redox.

The benefits of this are as follows:
- Easier to develop for Redox - `rustup target add x86_64-unknown-redox`
- Ensuring that changes to `libstd` also get propogated to `sys::redox`
- Improving Redox support further
- Eventually packaging `rustc` and `cargo` for Redox

**Comment by jackpot51** [comment]
@brson @alexcrichton @sfackler @BurntSushi @japaric I am curious about your opinions

**Comment by aturon** [comment]
cc @rust-lang/core for policy issue here.

Note that in general, we have several other outstanding requests for additional platform support, which are currently awaiting a general policy decision.

**Comment by sfackler** [comment]
If we did this, what level of effort would we need to put forward when implementing new APIs other than making sure the Redox build still compiles? For example, would it still be acceptable to stub out the implementation of `TcpStream::connect_timeout` with one that just returns an error?

**Comment by jackpot51** [comment]
Absolutely @sfackler. Returning ENOSYS for new API would be acceptable, but we would fill in the stubs as soon as possible.

**Comment by nikomatsakis** [comment]
@jackpot51 

> Due to recent advances in Redox, I think it is now time to promote it to a Tier 2 platform.

Forgive my ignorance, but I'm curious what advances you are referring to? (Just a link or two would suffice.)

**Comment by jackpot51** [comment]
The news articles since This Week in Redox 20, found here https://www.redox-os.org/news/ , sum it up

**Comment by aturon** [comment]
@jackpot51 I had a chance to talk with core team and infra folks about this question. In principle, we're open to moving platforms to tier 2 as long as there is solid evidence that community members will quickly address issues or blockages that arise -- and it's clear that Redox meets that bar.

However, at the moment, we are also at the limit of our build capacity. We're in the process of adding more capacity, but it will take some time. We should pick back up the discussion at that point.

To set clear expectations: in the long run, we want to move to a broader sponsorship funding model, so that we can support a wider array of platforms based on community support and demand. We're still working out the details and policies around that, though. In the short term, however, I expect we'll have enough extra capacity that we could do this for some time.

**Comment by jackpot51** [comment]
@aturon @sfackler @nikomatsakis @brson Here is an implementation: https://github.com/rust-lang/rust/pull/43303

**Comment by musoke** [comment]
Is this issue closed by #43303 ?

**Comment by jackpot51** [comment]
Yes, it is.
